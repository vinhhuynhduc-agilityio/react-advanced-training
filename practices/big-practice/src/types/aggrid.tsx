import { ColDef } from "ag-grid-community";
import { RowData } from "./users";

export interface DataGridProps {
  rowData: RowData[];
  columnDefs: ColDef<RowData>[];
};