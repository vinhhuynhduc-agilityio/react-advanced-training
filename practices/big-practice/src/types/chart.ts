
export interface FormattedDataTotalTasksCompleted {
  month: string;
  [year: number]: number;
};

export interface FormattedDataTotalTasksByProjects {
  projectName: string;
  [year: number]: number;
};
