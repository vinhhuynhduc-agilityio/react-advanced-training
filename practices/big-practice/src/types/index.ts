export * from './chart';
export * from './task';
export * from './user';
export * from './project';
