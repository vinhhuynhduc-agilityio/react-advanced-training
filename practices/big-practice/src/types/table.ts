export interface UserData {
  id: string;
  fullName: string;
  earnings: string;
  email: string;
  avatarUrl: string;
  registered: string;
  lastUpdated: string;
}

export interface TaskData {
  id: string;
  userId: string;
  projectId: string;
  taskName: string;
  startDate: string;
  completedDate: string;
  currency: number;
  status: boolean;
  projectName: string;
  fullName: string;
  tooltipShowDelay?: number;
}

export interface ProjectsData {
  id: string;
  projectName: string;
}

export interface UserListDrawerProps {
  users: UserData[];
  selectedUserId: string | null;
  onUserSelected: (userId: string | null) => void;
  sourceComponent: string | null;
}

export type StatusWithDate = {
  status: boolean;
  completedDate: string;
};

export type FieldValue =
  | string
  | number
  | boolean
  | ProjectsData
  | UserData
  | StatusWithDate;
