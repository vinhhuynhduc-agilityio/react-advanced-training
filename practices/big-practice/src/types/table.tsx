import {
  ColDef,
  GetRowIdParams,
  GridReadyEvent,
  RowClassParams,
  RowClickedEvent
} from "ag-grid-community";

export interface UserData {
  id: string;
  fullName: string;
  earnings: string;
  email: string;
  avatarUrl: string;
  registered: string;
  lastUpdated: string;
};

export interface TaskData {
  id: string;
  userId: string;
  projectId: string;
  taskName: string;
  startDate: string;
  completedDate: string;
  currency: number;
  status: boolean;
  projectName: string;
  fullName: string;
};

export interface ProjectsData {
  id: string;
  projectName: string;
};

export interface DataGridProps<T> {
  rowData: T[];
  userData?: T[];
  columnDefs: ColDef<T>[];
  onRowClicked?: (event: RowClickedEvent) => void;
  getRowClass?: (params: RowClassParams) => string;
  rowHeight?: number;
  onGridReady: (event: GridReadyEvent) => void;
  getRowId: (params: GetRowIdParams<T>) => string;
};

export interface UserListDrawerProps {
  users: UserData[];
  selectedUserId: string | null;
  onUserSelected: (userId: string | null) => void;
  sourceComponent: string | null;
};

export interface TaskDataProps {
  tasks: TaskData[];
  selectedUserId: string | null;
  onTaskRowSelected: (userId: string | null) => void;
  projects: ProjectsData[];
  sourceComponent: string | null;
  users: UserData[]
};
