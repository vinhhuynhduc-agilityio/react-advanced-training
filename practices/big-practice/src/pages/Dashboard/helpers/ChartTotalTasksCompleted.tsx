import { FormattedMonthData } from "@/types/chartTypes";
import { TaskData } from "@/types/table";

/**
 * @returns {FormattedMonthData[]} Array containing task completion data by month,
 * each object contains: `month` (month name), `2023` and `2024` (number of completed tasks)
 */
const formatDataForChartTotalTasks = (tasks: TaskData[]): FormattedMonthData[] => {
  const months = {
    Jan: { month: "Jan", 2023: 0, 2024: 0 },
    Feb: { month: "Feb", 2023: 0, 2024: 0 },
    Mar: { month: "Mar", 2023: 0, 2024: 0 },
    Apr: { month: "Apr", 2023: 0, 2024: 0 },
    May: { month: "May", 2023: 0, 2024: 0 },
    Jun: { month: "Jun", 2023: 0, 2024: 0 },
    Jul: { month: "Jul", 2023: 0, 2024: 0 },
    Aug: { month: "Aug", 2023: 0, 2024: 0 },
    Sep: { month: "Sep", 2023: 0, 2024: 0 },
    Oct: { month: "Oct", 2023: 0, 2024: 0 },
    Nov: { month: "Nov", 2023: 0, 2024: 0 },
    Dec: { month: "Dec", 2023: 0, 2024: 0 },
  };

  tasks.forEach((task) => {
    const completedDate = task.completedDate;

    if (completedDate !== "incomplete") {
      const [, monthStr, yearStr] = completedDate.split(" ");
      const year = parseInt("20" + yearStr);

      if (year === 2023 || year === 2024) {
        months[monthStr as keyof typeof months][year]++;
      }
    }
  });

  return Object.values(months);
};

export {
  formatDataForChartTotalTasks
};
