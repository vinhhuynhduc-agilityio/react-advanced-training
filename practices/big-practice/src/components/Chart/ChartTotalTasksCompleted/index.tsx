import {
  useMemo,
  memo
} from 'react';
import { AgCharts } from 'ag-charts-react';

// component
import { Spinner } from '@/components/common';

// helpers
import { formatDataForChartTotalTasks, totalTasksCompletedOptions } from '@/components/Chart/helpers';

// types
import { TaskData } from '@/types';

interface ChartTotalTasksCompletedProps {
  isLoading: boolean;
  tasks: TaskData[]
};

export const ChartTotalTasksCompleted: React.FC<ChartTotalTasksCompletedProps> = memo(
  ({ isLoading, tasks }) => {

    const formattedData = useMemo(() => formatDataForChartTotalTasks(tasks), [tasks]);

    const memoizedOptions = useMemo(() => ({
      ...totalTasksCompletedOptions,
      data: formattedData,
    }), [formattedData]);

    if (isLoading) {
      return (
        <div
          className="flex-1 bg-white border border-customBorder h-[302px]"
          role="figure"
          aria-label="Loading chart: Total tasks completed"
        >
          <Spinner />
        </div>
      );
    }

    return (
      <div
        id="chart-total-tasks-completed"
        role="figure"
        aria-label="Chart showing total tasks completed"
        className="flex-1 bg-white border border-customBorder h-[302px]"
      >
        <AgCharts options={memoizedOptions} />
      </div>
    );
  }
);
