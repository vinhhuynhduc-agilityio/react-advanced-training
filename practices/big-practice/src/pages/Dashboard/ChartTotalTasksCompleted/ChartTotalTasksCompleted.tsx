import {
  useState,
  useEffect,
  useMemo,
  memo
} from "react";
import { AgCharts } from "ag-charts-react";
import {
  AgBarSeriesTooltipRendererParams,
  AgChartOptions,
  AgTooltipRendererResult
} from "ag-charts-community";

// types
import { TaskData } from "@/types/table";

// helpers
import { formatDataForChartTotalTasks } from "../helpers/ChartTasks";

// component
import Spinner from "@/components/Spinner/Spinner";

interface ChartTotalTasksCompletedProps {
  tasks: TaskData[];
  isLoading: boolean;
};

const ChartTotalTasksCompleted: React.FC<ChartTotalTasksCompletedProps> = ({
  tasks,
  isLoading
}) => {
  const formattedData = useMemo(
    () => formatDataForChartTotalTasks(tasks),
    [tasks]
  );

  const renderTooltipChart = (params: AgBarSeriesTooltipRendererParams): AgTooltipRendererResult => {
    const month = params.datum[params.xKey];
    const tasksCompleted = params.datum[params.yKey];

    return {
      title: `<div style="text-align: center; line-height: 1.5">
                <div>${month}</div>
                <div>${tasksCompleted} tasks completed</div>
              </div>`,
      content: '',
      backgroundColor: '#181d1f'
    };
  };

  const [options, setOptions] = useState<AgChartOptions>({
    title: {
      text: "Total tasks completed",
    },
    data: formattedData,
    series: [
      {
        type: "line",
        xKey: "month",
        yKey: "2023",
        yName: "2023",
        tooltip: {
          renderer: renderTooltipChart
        },
      },
      {
        type: "line",
        xKey: "month",
        yKey: "2024",
        yName: "2024",
        tooltip: {
          renderer: renderTooltipChart
        },
      },
    ],
    legend: {
      enabled: true,
      position: 'bottom',
      item: {
        marker: {
          size: 10,
        },
        label: {
          fontWeight: 'bold',
        },
      },
    },
  });

  useEffect(() => {
    setOptions((prevOptions) => ({
      ...prevOptions,
      data: formattedData,
    }));
  }, [formattedData]);

  if (isLoading) {
    return (
      <div className="flex-1 bg-white border border-customBorder h-[302px]">
        <Spinner />
      </div>
    );
  }

  return (
    <div className="flex-1 bg-white border border-customBorder h-[302px]">
      <AgCharts options={options} />
    </div>
  )
};

export default memo(ChartTotalTasksCompleted);
