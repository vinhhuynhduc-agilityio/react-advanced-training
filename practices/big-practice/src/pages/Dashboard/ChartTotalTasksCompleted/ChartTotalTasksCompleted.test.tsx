import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { TaskData } from "@/types/table";

// component
import ChartTotalTasksCompleted from "./ChartTotalTasksCompleted";

// Mock AgCharts
jest.mock("ag-charts-react", () => ({
  AgCharts: () => <div data-testid="mocked-chart"></div>,
}));

// Mock helpers
jest.mock("../helpers/ChartTasks", () => ({
  ...jest.requireActual("../helpers/ChartTasks"),
  formatDataForChartTotalTasks: jest.fn(() => [
    { month: "January", "2023": 5, "2024": 10 },
    { month: "February", "2023": 3, "2024": 7 },
  ]),
}));

describe("ChartTotalTasksCompleted Component", () => {
  const mockTasks: TaskData[] = [
    {
      "id": "71e564f4-7c18-47f7-89f2-abe4b7ec2854",
      "userId": "d290f1ee-6c54-4b01-90e6-d701748f0851",
      "projectId": "f2d32cb6-12c7-4ae7-bb28-74f16d1d2cbb",
      "taskName": "Build test",
      "startDate": "10 Aug 24",
      "completedDate": "15 Nov 24",
      "currency": 2000,
      "status": true,
      "projectName": "Support",
      "fullName": "Joe Bloggs"
    },
  ];

  it("renders the mocked chart", () => {
    render(<ChartTotalTasksCompleted tasks={mockTasks} />);
    const mockedChart = screen.getByTestId("mocked-chart");
    expect(mockedChart).toBeInTheDocument();
  });

  it("matches snapshot for mocked chart", () => {
    const { container } = render(<ChartTotalTasksCompleted tasks={mockTasks} />);
    expect(container).toMatchSnapshot();
  });
});
