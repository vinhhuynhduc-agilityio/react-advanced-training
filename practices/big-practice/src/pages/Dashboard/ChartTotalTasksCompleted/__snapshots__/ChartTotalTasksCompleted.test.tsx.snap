// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChartTotalTasksCompleted Component matches snapshot for mocked chart 1`] = `
<div>
  <div
    class="flex-1 bg-white border border-customBorder h-[302px]"
  >
    <div
      data-testid="mocked-chart"
    />
  </div>
</div>
`;

exports[`ChartTotalTasksCompleted Component matches snapshot for rendered chart 1`] = `
<div>
  <div
    class="flex-1 bg-white border border-customBorder h-[302px]"
    data-testid="total-tasks-chart"
  >
    <div>
      <div
        class="ag-charts-wrapper ag-charts-styles"
        data-ag-charts=""
        data-animating="false"
        role="presentation"
        style="width: 300px; height: 300px;"
      >
        <div
          class="ag-charts-canvas-center"
          role="presentation"
        >
          <div
            class="ag-charts-canvas-container"
            role="presentation"
          >
            <div
              aria-hidden="true"
              class="ag-charts-canvas"
              role="presentation"
            >
              <canvas
                height="300"
                style="width: 600px; height: 300px;"
                width="600"
              />
            </div>
            <div
              class="ag-charts-canvas-proxy"
              role="figure"
            >
              <div
                aria-labelledby="SeriesAreaManager-2-aria-label"
                class="ag-charts-series-area"
                role="img"
                tabindex="-1"
              >
                <div
                  class="ag-charts-focus-indicator"
                >
                  <svg>
                    <path />
                  </svg>
                </div>
                <div
                  id="SeriesAreaManager-2-aria-label"
                  style="display: none;"
                />
              </div>
            </div>
            <div
              class="ag-charts-canvas-overlay"
              role="presentation"
            >
              <div
                class="ag-chart-tooltip"
              />
              <div
                class="ag-chart-overlay"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
